config/deprecation.go

```diff
@@ -0,0 +1,359 @@
+package config
+
+import (
+	"fmt"
+	"log" //nolint:revive // log is ok here as the logging facility is not set-up yet
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/fatih/color"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/aggregators"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/outputs"
+	"github.com/influxdata/telegraf/plugins/processors"
+)
+
+// Escalation level for the plugin or option
+type Escalation int
+
+func (e Escalation) String() string {
+	switch e {
+	case Warn:
+		return "WARN"
+	case Error:
+		return "ERROR"
+	}
+	return "NONE"
+}
+
+const (
+	// None means no deprecation
+	None Escalation = iota
+	// Warn means deprecated but still within the grace period
+	Warn
+	// Error means deprecated and beyond grace period
+	Error
+)
+
+// deprecationInfo contains all important information to describe a deprecated entity
+type deprecationInfo struct {
+	// Name of the plugin or plugin option
+	Name string
+	// Level of deprecation
+	Level Escalation
+	info  telegraf.DeprecationInfo
+}
+
+func (di *deprecationInfo) determineEscalation(version *semver.Version) {
+	di.Level = None
+	if di.info.Since == "" {
+		return
+	}
+
+	since, err := semver.NewVersion(di.info.Since + ".0")
+	if err != nil {
+		panic(fmt.Errorf("cannot parse 'since' version %q: %v", di.info.Since, err))
+	}
+
+	var removal *semver.Version
+	if di.info.RemovalIn != "" {
+		removal, err = semver.NewVersion(di.info.RemovalIn + ".0")
+		if err != nil {
+			panic(fmt.Errorf("cannot parse 'removal' version %q: %v", di.info.RemovalIn, err))
+		}
+	} else {
+		removal = &semver.Version{Major: since.Major, Minor: since.Minor}
+		removal.BumpMajor()
+		di.info.RemovalIn = strings.TrimSuffix(removal.String(), ".0")
+	}
+
+	if !version.LessThan(*removal) {
+		di.Level = Error
+	} else if !version.LessThan(*since) {
+		di.Level = Warn
+	}
+}
```

Comment:
Something here is also not working:

```
❯ cat build_version.txt 
2.0.0
telegraf on  deprecation [!] via 🐹 v1.17.3 
❯ make
go mod download -x
go build -ldflags " -X main.commit=6132f539 -X main.branch=deprecation -X main.goos=linux -X main.goarch=amd64 -X main.version=2.0.0-6132f539" ./cmd/telegraf
telegraf on  deprecation [!] via 🐹 v1.17.3 took 3s 
❯ ./telegraf --config config.toml --debug
2021-11-15T21:58:59Z I! Starting Telegraf 2.0.0-6132f539
2021-11-15T21:58:59Z W! DeprecationWarning: Plugin "inputs.io" deprecated since version 0.10 and will be removed in 2.0: use 'inputs.diskio' instead
```

However, as soon as the patch or minor version gets increments, it works:

```
./telegraf --config config.toml --debug
2021-11-15T22:03:31Z I! Starting Telegraf 2.0.1-6132f539
2021-11-15T22:03:31Z E! DeprecationError: Plugin "inputs.io" deprecated since version 0.10 and will be removed in 2.0: use 'inputs.diskio' instead
2021-11-15T22:03:31Z E! [telegraf] Error running agent: Error loading config file config.toml: error parsing io, plugin deprecated
```

Reply:
I think it works as intended. In my test I have
```shell
2021-11-16T10:46:45Z I! Starting Telegraf 2.0.0-17dd4743
version = 2.0.0-17dd4743 (PreRelease=17dd4743)
```
which means that we are **pre** version 2.0.0 due to the `-17dd4743` pre-release tag. As soon as you use `2.0.0` instead of `2.0.0-something` the warning will turn into an error.

So there are two ways to deal with it

1. Keep as is and realize the deprecation _with the release_.
1. We "overwrite" the pre-release version for comparison and check only based on major, minor and path.

Which one do you prefer?

Reply:
ahh right, the pre-release identifier was throwing it off. Thanks for catching that. I think tossing the pre-release identifier during the comparison is the right thing to do. That way people who are trying the next version get the real behavior. So option 2 :)

Thoughts?

Reply:
Sounds reasonable..